dlhub_sdk/utils/validation.py

```diff
@@ -0,0 +1,239 @@
+        raise _generate_err(ValueError, path, msg=f"dl.run expected tuple of length {len(db_entries)}, recieved tuple with length {len(tup)}"
+                                                  + "{loc}")
+
+    path.append(["tuple", None])
+    for i, (given, expected) in enumerate(zip(tup, db_entries)):
+        path[-1][1] = i
+        validate(given, expected, path)
+    path.pop()
+
+
+def _validate_ndarray(arr: ndarray, shape: list, db_entry: dict, path: list) -> None:
+    """Compare the shape of arr with shape and recursively validate each of the elements in arr against db_entry["item_type"]
+
+    Args:
+        arr (ndarray): The ndarray whose shape and items need to be validated
+        shape (list): The shape that arr is expected to have
+        db_entry (dict): The entry where type data may or may not be retrieved
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If arr.shape does not match shape
+        TypeError: If any of the items' types do not match
+    """
+    shape = tuple(int(x) if x != "None" else x for x in shape)
+    shape_err = _generate_err(ValueError, path, msg=f"dl.run given improper input: expected ndarray.shape = {shape}, received shape: {arr.shape}"
+                                                    + "{loc}")
+
+    if len(arr.shape) != len(shape):
+        raise shape_err
+    if not all(map(lambda t: t[0] == t[1] or t[1] == "None", zip(arr.shape, shape))):
+        raise shape_err
+
+    entry = db_entry.get("item_type")
+    if entry:
+        path.append(["ndarray", None])
+        for i, item in ndenumerate(arr):
+            path[-1][1] = i
+            validate(item.item(), entry, path)
+        path.pop()
+
+
+def _validate_dict(dct: dict, props: dict, path: list) -> None:
+    """Recursively validate each of the pairs in dct against props
+
+    Args:
+        dct (dict): The dict whose keys and values need to be validated
+        props (dict): The dict that describes the expected keys and values
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If dct is missing a key from props or a value in dct causes an error
+        TypeError: If any of the values' types do not match
+    """
+    if props:
+        path.append(["dict", None])
+        for key in dct:
+            path[-1][1] = key
+            if key not in props:
+                raise _generate_err(ValueError, path, msg=f"dl.run given improper input: given unexpected dictionary key: {repr(key)}"+"{loc}")
+        for key in props:
+            path[-1][1] = key
+            if key not in dct:
+                raise _generate_err(ValueError, path, msg=f"dl.run given improper input: expected dictionary key: {repr(key)} to be present"+"{loc}")
+            validate(dct[key], props[key], path)
+        path.pop()
+
+
+def _test() -> None:
```

Comment:
Tests should go in a test file, not in the file with the code.

Reply:
Check out https://docs.pytest.org/en/7.1.x/ for how to write tests. Just add another file in https://github.com/DLHub-Argonne/dlhub_sdk/blob/master/dlhub_sdk/utils/tests/

Really, all you need to do is move this test function into a file named `test_validation.py` and remove the underscore from the test name.

Reply:
The tests look fine, but running them automatically in GHA will tell us if you're missing anything